Split deploy canary demo script into smaller scripts

harshitOberoi opened the pull request:
**Issue #:**
After adding deployment of ChimeSDKMeetings specific demos to `Deploy Canary Demo` GitHub Action which works on merging into master branch the average deployment time increased from 25 minutes to 1 hour.

**Description of changes:**
1) Removed `deploy-canary-demo` script which builds browser demo and deploys serveless demo to all the canary regions and split it into region specific scripts:
     a) build-browser-demo 
     b) deploy-canary-demo-alpha
     c) deploy-canary-demo-beta
     d) deploy-canary-demo-gamma
2) Updated `deploy.yml` to add multiple jobs to run all the above scripts in parallel

**Testing:**

*Can these tested using a demo application? Please provide reproducible step-by-step instructions.*
No


**Checklist:**

1. Have you successfully run `npm run build:release` locally?
Yes


2. Do you add, modify, or delete public API definitions? If yes, has that been reviewed and approved?


3. Do you change the wire protocol, e.g. the request method? If yes, has that been reviewed and approved?


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

devalevenkatesh commented on .github/workflows/deploy.yml:
```diff
@@ -8,12 +8,56 @@ on:
+      - name: Install SAM CLI
+        run: script/github-action-awscli-installation
+      - name: Build Browser Demo
+        run: script/build-browser-demo
```
We can just add steps here? It is just two steps so fine to not to add into its own script.

ltrung replied:
Agree

harshitOberoi replied:
made the change

devalevenkatesh commented on .github/workflows/deploy.yml:
```diff
@@ -8,12 +8,56 @@ on:
+        run: script/github-action-awscli-installation
+      - name: Build Browser Demo
+        run: script/build-browser-demo
+  deploy_alpha_canary:
```
Each of the deploy* job needs to have a "needs" attribute for the `build_browser_demo:` job to finish before the deploy job runs.

Reference: https://github.com/aws/amazon-chime-sdk-js/blob/master/.github/workflows/browser-compatibility-test.yml#L122

ltrung replied:
This build browser demo have to be repeated since each of these job has its own VM. You build the demo in one VM but the publish is another. Also we do not need to build the demo anymore. The deploy script will automatically build the demo for you.

harshitOberoi replied:
@ltrung  The statement is contradicting itself. I thought the script might just be testing that browser demo is building correctly.

harshitOberoi replied:
@devalevenkatesh I followed this workflow as there are two jobs but one does not depend on other https://github.com/aws/amazon-chime-sdk-js/blob/SplitDeployCanaryDemo/.github/workflows/webview-ui-test.yml

harshitOberoi replied:
Also if I put needs these jobs will run in sequential order not parallely.

harshitOberoi replied:
Tested it out jobs do not need to depend on each other

devalevenkatesh commented on .github/workflows/deploy.yml:
```diff
@@ -41,4 +81,27 @@ jobs:
+      - name: Configure AWS Credentials
+        uses: aws-actions/configure-aws-credentials@v1
+        with:
+          aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY }}
```
Should these use `${{secrets.GAMMA_AWS_ACCESS_KEY}}` and `${{secrets.GAMMA_AWS_SECRET_ACCESS_KEY}}`?

devalevenkatesh replied:
Similarly for the beta case as well.

harshitOberoi replied:
So the previous script was also not using these but still present i was not sure what they were for so kept it will remove it now.

harshitOberoi replied:
I am not sure which credentials we should used for running the script.

michhyun1 replied:
```
echo "Deploying to beta stage for canary that talks to gamma Chime endpoint"
export AWS_ACCESS_KEY_ID=$BETA_AWS_ACCESS_KEY_ID
export AWS_SECRET_ACCESS_KEY=$BETA_AWS_SECRET_ACCESS_KEY
```

I think the deploy-canary-demo script exports these env variables before running the deploy script. I wonder if we need this step at all.

devalevenkatesh commented on .github/workflows/deploy.yml:
```diff
@@ -8,12 +8,56 @@ on:
+  build_browser_demo:
+    name: Build Browser Demo
+    runs-on: ubuntu-latest
+    env:
```
For `build_browser_demo` we dont need this section and this AWS configuring should go into each deploy job?
```
env:
        AWS_DEFAULT_REGION: us-east-1
        AWS_DEFAULT_OUTPUT: text
        AWS_ACCESS_KEY_ID: ${{secrets.AWS_ACCESS_KEY}}
        AWS_SECRET_ACCESS_KEY: ${{secrets.AWS_SECRET_ACCESS_KEY}}
    steps:
      - name: Configure AWS Credentials
        uses: aws-actions/configure-aws-credentials@v1
        with:
          aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY }}
          aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
          aws-region: us-east-1
      - name: Checkout Package
        uses: actions/checkout@v2
        with:
          fetch-depth: 0
      - name: Install SAM CLI
        run: script/github-action-awscli-installation
```

michhyun1 replied:
Yeah, each job runs in its own container with no context from the other jobs, so you can't build it once here and expect the other jobs to have the meeting and meeting readiness checker built. 

If you want to build the demo before deploying them, you'll have to build them in each job seperately. Else, you can do `npm run build` within the deploy script instead of having it in the github action. That might be better.

harshitOberoi replied:
Tested we do not need to run "npm run build" deploy script builds the app before deploying

ltrung commented on .github/workflows/deploy.yml:
```diff
@@ -8,12 +8,56 @@ on:
+      - name: Build Browser Demo
+        run: script/build-browser-demo
+  deploy_alpha_canary:
+    name: Deploy Demo App on Alpha Canary
```
there is no such thing as alpha canary :D Maybe we should rename it to something else.

harshitOberoi replied:
Renamed

ltrung commented on script/deploy-canary-demo-alpha:
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+set -euxo pipefail
```
Do we need these script or can we just put them under the github action itself

harshitOberoi replied:
Currently all the github actions calls the scripts so i was following the same pattern but let me know if you want to remove all these scrtips and just add it inside job steps

ltrung replied:
Previously the script is pretty long so we just dump all commands into a file since you split them all out they are not that long anymore so I think we can just put that in github action.

harshitOberoi replied:
@ltrung I tried but not able to run commands directly

harshitOberoi replied:
@ltrung Updated such that all the jobs use a single script file and added a new env variable "NAME" based upon which script deploys correct demos.

michhyun1 commented on .github/workflows/deploy.yml:
```diff
@@ -8,12 +8,60 @@ on:
+      - name: Install SAM CLI
+        run: script/github-action-awscli-installation
+      - name: Deploy the Serverless Demo App
+        run: script/deploy-stanup-demo
```
typo?

harshitOberoi replied:
fixed

michhyun1 commented on .github/workflows/deploy.yml:
```diff
@@ -41,4 +85,27 @@ jobs:
       - name: Install SAM CLI
```
Dont we need to use the beta aws access and security key above?

harshitOberoi replied:
fixed

michhyun1 commented on script/deploy-stanup-demo:
```diff
@@ -0,0 +1,10 @@
+
+cd $GITHUB_WORKSPACE/demos/serverless
+npm ci
+
```
Maybe we can export the main aws creds here, the same way we do for the deploy gamma and deploy beta. We might not need the configure AWS creds step in that case I think.

harshitOberoi replied:
fixed

michhyun1 commented on .github/workflows/deploy.yml:
```diff
@@ -8,12 +8,60 @@ on:
+        run: npm run build
+      - name: Build Meeting Readiness Checker
+        run: npm run build --app=meetingReadinessChecker
+  deploy_standup_demo:
```
deploy alpha demo

michhyun1 commented on .github/workflows/deploy.yml:
```diff
@@ -8,12 +8,60 @@ on:
+        run: script/github-action-awscli-installation
+      - name: Deploy the Serverless Demo App
+        run: script/deploy-stanup-demo
+  deploy_beta_canary:
```
deploy beta demo

michhyun1 commented on .github/workflows/deploy.yml:
```diff
@@ -41,4 +85,27 @@ jobs:
       - name: Deploy the Serverless Demo App
-        run: script/deploy-canary-demo
+        run: script/deploy-canary-demo-beta
+  deploy_gamma_canary:
```
deploy_gamma_demo

michhyun1 commented:
I think you have the right idea - I think you want 3 jobs:

1. deploy_alpha_demo
- checkout package
- run alpha demo deploy script (cd into demo, npm i, npm build, set alpha aws creds, deply)

2. deploy_beta_demo
- checkout package
- run beta demo deploy script (cd into demo, npm i, npm build, set beta aws creds, deply)

3. deploy_gamma_demo
- checkout package
- run gamma demo deploy script (cd into demo, npm i, npm build, set gamma aws creds, deply)

ltrung commented on script/deploy-chime-alpha-demo:
```diff
@@ -0,0 +1,10 @@
+
+echo "Deploying to alpha stage for canary"
+
+npm run deploy -- -b chime-sdk-demo-canary -s chime-sdk-demo-canary -o chime-sdk-demo-canary -u false -i eu-south-1 -t -l
```
If the script is this short, there is no reason to have separate script, you can just put stuff straight to GitHub action

harshitOberoi replied:
I tried it a lot of times everytime deploy command was giving some issue when i tried to run it directly in github action. All the failed attempts https://github.com/aws/amazon-chime-sdk-js/actions/workflows/deploy.yml

ltrung replied:
that just seems you are in the wrong folder no?

michhyun1 replied:
Is there any use case of us wanting to use this besides the github action? manual deployment? If so, we want to have it as a seperate script @ltrung

ltrung replied:
No these scripts are only used for github actions

harshitOberoi replied:
I tried multiple time to cd into serverless folder but that command was not working.

harshitOberoi replied:
@ltrung Updated such that all the jobs use a single script file and added a new env variable "NAME" based upon which script deploys correct demos.

harshitOberoi commented on .github/workflows/deploy.yml:
```diff
@@ -1,44 +1,140 @@
 jobs:
-  deploy:
-    name: Deploy Demo App
+  build_browser_demos:
```
As some of the canaries use browser demo. Want to verify browser demo is also building correctly and also this step was already present in the original script.

ltrung replied:
No reason to do so. If it fails then anyof the deploy will fail anyway. That step is no longer necessary after our recent update.

michhyun1 replied:
yeah I agree we dont need this job.

harshitOberoi replied:
removed

ltrung reviewed (commented):
Do you commit the logout file by mistake?

ltrung commented on .github/workflows/deploy.yml:
```diff
@@ -1,44 +1,126 @@
 # For more information see: https://help.github.com/actions/language-and-framework-guides/using-nodejs-with-github-actions
 
-name: Deploy Demo App Workflow
+name: Deploy Meeting And Meeting Readiness Demos
```
nit: Can we keep the previous name as we might add more demos in the future :)?

harshitOberoi replied:
fixed

ltrung commented on script/deploy-canary-demo:
```diff
@@ -1,44 +1,49 @@
 cd $GITHUB_WORKSPACE/demos/serverless
 npm ci
 
+if [ $NAME == ChimeAlpha ]
```
nit: Can we do a case statement ?

harshitOberoi replied:
Fixed

harshitOberoi commented:
@ltrung lockfile was added as i had to provide executable access to deploy_canary_demo script. As i removed and then added it again

ltrung commented on .github/workflows/deploy.yml:
```diff
@@ -8,37 +8,119 @@ on:
+    env:
+        AWS_DEFAULT_REGION: us-east-1
+        AWS_DEFAULT_OUTPUT: text
+        STAGE: ChimeBeta
```
Shouldn't this be just `Beta` to be inconsistent with `Alpha` like above? Same for below stage?

harshitOberoi replied:
Alpha does not have two clients that is why i just put alpha and beta and gamma has two clients "Chime" and "ChimeMeetings"

ltrung commented on .github/workflows/deploy.yml:
```diff
@@ -8,37 +8,119 @@ on:
+    env:
+        AWS_DEFAULT_REGION: us-east-1
+        AWS_DEFAULT_OUTPUT: text
+        STAGE: ChimeMeetingsBeta
```
This isn't very clear. Can we name it to `RegionalizedBeta` or something?

harshitOberoi replied:
No we are refraining from using regional for ChimeMeetings client that is why i am using chime and chimeMeetings to differentiate between clients

devalevenkatesh commented on .github/workflows/deploy.yml:
```diff
@@ -8,37 +8,119 @@ on:
+      - name: Install SAM CLI
+        run: script/github-action-awscli-installation
+      - name: Run Deployment Script
+        run: script/deploy-canary-demo
```
I found running `script/deploy-canary-demo` under each job and then handling a switch based on `NAME` environment variable inside this script seems counter intuitive to me.

Would it be better to just drop that script file and run the commands as part of the `run` itself?

For example:
```
name: Deploy Demo App Workflow

on:
  push:
    branches: [ master ]

jobs:
  deploy_chime_alpha_demo:
    name: Alpha - Deploy Meeting Demo And Meeting Readiness Checker
    runs-on: ubuntu-latest
    env:
        AWS_DEFAULT_REGION: us-east-1
        AWS_DEFAULT_OUTPUT: text
        NAME: ChimeAlpha
    steps:
      - name: Configure AWS Credentials
        uses: aws-actions/configure-aws-credentials@v1
        with:
          aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY }}
          aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
          aws-region: us-east-1
      - name: Checkout Package
        uses: actions/checkout@v2
        with:
          fetch-depth: 0
      - name: Install SAM CLI
        run: script/github-action-awscli-installation
      - name: Run Deployment Script
        run:  |
          echo "Deploying to alpha stage for canary"
          npm run deploy -- -b chime-sdk-demo-canary -s chime-sdk-demo-canary -o chime-sdk-demo-canary -u false -i eu-south-1 -t -l
          npm run deploy -- -b chime-sdk-meeting-readiness-checker-dev-canary -s chime-sdk-meeting-readiness-checker-dev-canary -a meetingReadinessChecker -u false -t -l
  
  deploy_chime_beta_demo:
    name: Beta - Deploy the Serverless Meeting and Meeting Readiness Checker Demos that talks to Chime Client
    runs-on: ubuntu-latest
    env:
        AWS_DEFAULT_REGION: us-east-1
        AWS_DEFAULT_OUTPUT: text
        NAME: ChimeBeta
        GAMMA_CHIME_ENDPOINT: ${{secrets.GAMMA_CHIME_ENDPOINT}}
        GAMMA_CHIME_SERVICE_PRINCIPAL: ${{secrets.GAMMA_CHIME_SERVICE_PRINCIPAL}}
    steps:
      - name: Configure AWS Credentials
        uses: aws-actions/configure-aws-credentials@v1
        with:
          aws-access-key-id: ${{ secrets.BETA_AWS_ACCESS_KEY }}
          aws-secret-access-key: ${{ secrets.BETA_AWS_SECRET_ACCESS_KEY }}
          aws-region: us-east-1
      - name: Checkout Package
        uses: actions/checkout@v2
        with:
          fetch-depth: 0
      - name: Install SAM CLI
        run: script/github-action-awscli-installation
      - name: Run Deployment Script
        run: |
          echo "Deploying to beta stage for canary that talks to gamma Chime client"
          # Uses Chime Client
          npm run deploy -- -b chime-sdk-demo-beta-canary -s chime-sdk-demo-beta-canary -o chime-sdk-demo-beta-canary -i eu-south-1 -p $GAMMA_CHIME_SERVICE_PRINCIPAL -c $GAMMA_CHIME_ENDPOINT -u false -t -l
          npm run deploy -- -b chime-sdk-meeting-readiness-checker-beta-canary -s chime-sdk-meeting-readiness-checker-beta-canary -a meetingReadinessChecker -c $GAMMA_CHIME_ENDPOINT -u false -t -l
  
  deploy_chime_sdk_meetings_beta_demo:
    name: Beta - Deploy the Serverless Meeting Demos that talks to ChimeSDKMeetings Client
    runs-on: ubuntu-latest
    env:
        AWS_DEFAULT_REGION: us-east-1
        AWS_DEFAULT_OUTPUT: text
        NAME: ChimeSDKMeetingsBeta
        GAMMA_CHIME_ENDPOINT_US_EAST_1: ${{secrets.GAMMA_CHIME_ENDPOINT_US_EAST_1}}
        GAMMA_CHIME_ENDPOINT_EU_CENTRAL_1: ${{secrets.GAMMA_CHIME_ENDPOINT_EU_CENTRAL_1}}
        GAMMA_CHIME_ENDPOINT_AP_SOUTHEAST_1: ${{secrets.GAMMA_CHIME_ENDPOINT_AP_SOUTHEAST_1}}
        GAMMA_CHIME_SERVICE_PRINCIPAL: ${{secrets.GAMMA_CHIME_SERVICE_PRINCIPAL}}
    steps:
      - name: Configure AWS Credentials
        uses: aws-actions/configure-aws-credentials@v1
        with:
          aws-access-key-id: ${{ secrets.BETA_AWS_ACCESS_KEY }}
          aws-secret-access-key: ${{ secrets.BETA_AWS_SECRET_ACCESS_KEY }}
          aws-region: us-east-1
      - name: Checkout Package
        uses: actions/checkout@v2
        with:
          fetch-depth: 0
      - name: Install SAM CLI
        run: script/github-action-awscli-installation
      - name: Run Deployment Script
        run:  |
          echo "Deploying to beta stage for canary that talks to gamma ChimeSDKMeetings client"
          # Uses ChimeSDKMeetings client
          npm run deploy -- -r us-east-1 -b chime-sdk-meetings-demo-beta-canary-us-east-1 -s chime-sdk-meetings-demo-beta-canary-us-east-1 -p $GAMMA_CHIME_SERVICE_PRINCIPAL -c $GAMMA_CHIME_ENDPOINT_US_EAST_1 -t -l
          npm run deploy -- -r eu-central-1 -b chime-sdk-meetings-demo-beta-canary-eu-central-1 -s chime-sdk-meetings-demo-beta-canary-eu-central-1 -p $GAMMA_CHIME_SERVICE_PRINCIPAL -c $GAMMA_CHIME_ENDPOINT_EU_CENTRAL_1 -t -l
          npm run deploy -- -r ap-southeast-1 -b chime-sdk-meetings-demo-beta-canary-ap-southeast-1 -s chime-sdk-meetings-demo-beta-canary-ap-southeast-1 -p $GAMMA_CHIME_SERVICE_PRINCIPAL -c $GAMMA_CHIME_ENDPOINT_AP_SOUTHEAST_1 -t -l
  
  deploy_chime_gamma_demo:
    name: Gamma - Deploy the Serverless Meeting and Meeting Readiness Checker Demos that talks to Chime Client
    runs-on: ubuntu-latest
    env:
        AWS_DEFAULT_REGION: us-east-1
        AWS_DEFAULT_OUTPUT: text
        NAME: ChimeGamma
    steps:
      - name: Configure AWS Credentials
        uses: aws-actions/configure-aws-credentials@v1
        with:
          aws-access-key-id: ${{ secrets.GAMMA_AWS_ACCESS_KEY }}
          aws-secret-access-key: ${{ secrets.GAMMA_AWS_SECRET_ACCESS_KEY }}
          aws-region: us-east-1
      - name: Checkout Package
        uses: actions/checkout@v2
        with:
          fetch-depth: 0
      - name: Install SAM CLI
        run: script/github-action-awscli-installation
      - name: Run Deployment Script
        run:  |
          echo "Deploying to gamma stage for canary that talks to prod Chime client"
          # Uses Chime Client
          npm run deploy -- -b chime-sdk-demo-gamma-canary -s chime-sdk-demo-gamma-canary -o chime-sdk-demo-gamma-canary -u false -i eu-south-1 -t -l
          npm run deploy -- -b chime-sdk-meeting-readiness-checker-gamma-canary -s chime-sdk-meeting-readiness-checker-gamma-canary -a meetingReadinessChecker -u false -t -l
  
  deploy_chime_sdk_meetings_gamma_demo:
    name: Gamma - Deploy the Serverless Meeting Demos that talks to ChimeSDKMeetings Client
    runs-on: ubuntu-latest
    env:
        AWS_DEFAULT_REGION: us-east-1
        AWS_DEFAULT_OUTPUT: text
        NAME: ChimeSDKMeetingsGamma
    steps:
      - name: Configure AWS Credentials
        uses: aws-actions/configure-aws-credentials@v1
        with:
          aws-access-key-id: ${{ secrets.GAMMA_AWS_ACCESS_KEY }}
          aws-secret-access-key: ${{ secrets.GAMMA_AWS_SECRET_ACCESS_KEY }}
          aws-region: us-east-1
      - name: Checkout Package
        uses: actions/checkout@v2
        with:
          fetch-depth: 0
      - name: Install SAM CLI
        run: script/github-action-awscli-installation
      - name: Run Deployment Script
        run:  |
          echo "Deploying to gamma stage for canary that talks to prod ChimeSDKMeetings client"
          # Uses ChimeSDKMeetings client
          npm run deploy -- -r us-east-1 -b chime-sdk-meetings-demo-gamma-canary-us-east-1 -s chime-sdk-meetings-demo-gamma-canary-us-east-1 -t -l
          npm run deploy -- -r us-west-2 -b chime-sdk-meetings-demo-gamma-canary-us-west-2 -s chime-sdk-meetings-demo-gamma-canary-us-west-2 -t -l
          npm run deploy -- -r eu-central-1 -b chime-sdk-meetings-demo-gamma-canary-eu-central-1 -s chime-sdk-meetings-demo-gamma-canary-eu-central-1 -t -l
          npm run deploy -- -r ap-southeast-1 -b chime-sdk-meetings-demo-gamma-canary-ap-southeast-1 -s chime-sdk-meetings-demo-gamma-canary-ap-southeast-1 -t -l
```

@ltrung could you provide any suggestion?

devalevenkatesh replied:
I discussed with Trung and found that you faced some issue with the script running like this and hence was split out into one. Please ignore this one.

devalevenkatesh commented on script/deploy-canary-demo:
```diff
@@ -1,44 +1,51 @@
 
-npm run deploy -- -b chime-sdk-demo-canary -s chime-sdk-demo-canary -o chime-sdk-demo-canary -u false -i eu-south-1 -t -l
-npm run deploy -- -b chime-sdk-meeting-readiness-checker-dev-canary -s chime-sdk-meeting-readiness-checker-dev-canary -a meetingReadinessChecker -u false -t -l
+    npm run deploy -- -b chime-sdk-demo-canary -s chime-sdk-demo-canary -o chime-sdk-demo-canary -u false -i eu-south-1 -t -l
```
Add #Uses Chime Client comment here as well?

harshitOberoi replied:
fixed

devalevenkatesh commented on .github/workflows/deploy.yml:
```diff
@@ -8,37 +8,119 @@ on:
-  deploy:
-    name: Deploy Demo App
+  deploy_chime_alpha_demo:
+    name: Alpha - Deploy Meeting Demo And Meeting Readiness Checker
```
Saw Trung's comment on keeping the name same so can we keep it like: `name: Alpha - Deploy Demo App`` similarly elsewhere.

michhyun1 replied:
Yeah, technically we deploy two demos here - no harm in having a more detailed title right?

ltrung reviewed (changes requested):
Is the logout.js committed by mistake?

harshitOberoi commented:
@ltrung  The logout.js was added as i deleted the deploy_demo_canary script and when i added it back i had to provide executable access and that added logout.js

michhyun1 commented:
Need to rebase and then force push again. LGTM otherwise - @ltrung @devalevenkatesh any concerns?

devalevenkatesh commented on .github/workflows/deploy.yml:
```diff
@@ -8,37 +8,119 @@ on:
-  deploy:
-    name: Deploy Demo App
+  deploy_chime_alpha_demo:
+    name: Alpha - Deploy Meeting Demo And Meeting Readiness Checker
```
Minor one, but checking this is not readable until you extend 😅 :
![image](https://user-images.githubusercontent.com/61809205/146835292-f9fb9c4a-e66a-47bb-adcd-34f8391bebe6.png)

Suggest to update the name with some differentiator:
Beta - Chime Client -Deploy Meeting and Meeting Readiness Checker Demos
Beta - ChimeSDKMeetings Client -Deploy Meeting and Meeting Readiness Checker Demos

harshitOberoi replied:
updated

ltrung reviewed (commented):
Do you need prod canary? If so should you update the postrelease script which release prod demo app.

harshitOberoi commented:
we do not need prod canaries as Gamma Canaries are testing ChimeMeetingSDK prod endpoints
